Stop using `pip` as an API

leorochael opened the pull request:
As of `pip` >= 10, internal APIs have been moved to discourage the kind
of usage we were doing.

Replace all pip calls with pip subprocess calls.

Replace all pip internal APIs with `pkg_resource` APIs.

Fixes #28

leorochael committed:
Simplify Python2/3 compatibility

Use `subprocess` throughout instead of `commands`.

Use stub functions instead of conditional checks.

leorochael committed:
Stop using `pip` as an API

As of `pip` >= 10, internal APIs have been moved to discourage the kind
of usage we were doing.

Replace all pip calls with pip subprocess calls.

Replace all pip internal APIs with `pkg_resource` APIs

leorochael commented:
@gershwinlabs, opinions on this patch?

billallen256 commented:
Please see my comments in #28.

billallen256 commented:
I have pulled this into the issue28 branch for further testing.

billallen256 commented:
Initial tests look promising. A couple of adjustments though...

We’ll need to stream the stderr/stdout from the pip call with Popen, probably something like this: https://stackoverflow.com/questions/2804543/read-subprocess-stdout-line-by-line

Still trying to figure out the best way to call the right pip as well. Calling pip from Python 3 doesn’t do any good if “pip” is associated with Python 2. Likewise for “pip3” from Py2.

leorochael commented:
I know how to locate the appropriate pip. It's just a matter of locating the appropriate Python. will adjust accordingly.

As for streaming, what is the exact effect you want? And why is the current handling of stderr/stdout in the PR wrong?

I though that by using `check_call()`, the standard setting of `{stdout,stderr}=None` should make the subprocess inherit the same file descriptors of the current process, which would have the same result as calling `pip` as a library.

Can you give me an example of output that is (not) showing that shouldn't (should)?

leorochael committed:
fixup! Stop using `pip` as an API

Make sure we execute the right pip from the right python, regardless of
which python/pip is in the path.

billallen256 commented:
The desired behavior is that stdout/stderr output from pip should display in real-time in the Jupyter notebook output. With check_call it does not do this, it just displays “0” when the command completes. This differs from the Python interactive prompt, where the output does display in real-time. I have not had a chance to fully explore the reason for this difference, but using subprocess.PIPE then iterating through the output lines (see stackoverflow link above) does have the desired bahvior in Jupyter.

leorochael commented:
> The desired behavior is that stdout/stderr output from pip should display in real-time in the Jupyter notebook output.

Thanks, I can try to work with that. Seems like Jupyter is capturing print output by setting `sys.std{out,err}` to file-like objects, but subprocesses ignore the `sys.std{out,err}` objects and inherit the file descriptors directly at the `C` level.

The IPython kernel in Jupyter does something a little more involved when you use the `!` magic command so as to properly capture output, as in:

![image](https://user-images.githubusercontent.com/239125/39204850-627b4432-47cf-11e8-98de-c5213c76ff34.png)

It's [using 'pexpect` to capture output](https://github.com/ipython/ipython/blob/5b2b7dd07a268baceeeedfe919de0a59e5bc922b/IPython/utils/_process_posix.py#L153) without blocking.

We could do the same in an `IPython` environment by calling `get_ipython().system()` (or, more directly, `IPython.utils.process.system()`), which is exactly what the `!` command in `IPython` is.

But this would imply a dependency on `IPython`, at least an optional one. We would also need to convert a list of args for pip into a single string embedding space separators and quoting appropriately.

Redirecting `stderr` to `STDOUT`, and `stdout` to `PIPE' on the subprocess then reading line by line, as mentioned in the SO answer you linked, lets us handle the output, including reporting it.

If we wanted to make sure that both `sys.stdout` and `sys.stderr` are "honored" in subprocess, then we would need to do something more involved, like reading both stdout and stderr from the subprocess in a loop with [`select()`](https://docs.python.org/3/library/select.html#select.select).

billallen256 commented:
I was going to opt for `select()`, but based on the docs it looks like that would break Windows compatibility.

An optional dependency on IPython isn't such a bad thing.  It would be nice if something even cleaner could be devised for IPython users though, where the success/failure of each package installation got printed out, one per line, and the user could expand to see the output if the installation failed.  I think iruby-dependencies does something like this.

Right now ipydeps takes all the desired packages and just passes them to pip so pip can work out the dependencies/order that things should be installed.  Calling pip for each package and getting the return status would be one way to get installation success/failure, but I'm not sure if this would actually work.  I'm not sure how smartly pip determines the installation order.  There was a related discussion in #13.
